qiskit/algorithms/fidelities/base_fidelity.py

```diff
@@ -0,0 +1,253 @@
+
+                right_parameters = ParameterVector("y", circuit_2.num_parameters)
+                self._right_parameters.append(right_parameters)
+                parametrized_circuit_2 = circuit_2.assign_parameters(right_parameters)
+
+                circuit = self._create_fidelity_circuit(
+                    parametrized_circuit_1, parametrized_circuit_2
+                )
+                circuits.append(circuit)
+                # update cache
+                self._circuit_cache[id(circuit_1), id(circuit_2)] = circuit
+
+        # set circuits
+        self._circuits = circuits
+
+    def _set_values(
+        self, values_1: Sequence[Sequence[float]] | None, values_2: Sequence[Sequence[float]] | None
+    ) -> None:
+        """
+        Update the list of parameter values to evaluate the corresponding
+        fidelity circuits with.
+
+        Args:
+            values_1: Numerical parameters to be bound to the first circuits.
+            values_2: Numerical parameters to be bound to the second circuits.
+
+        Raises:
+            ValueError: if the length of the input value lists doesn't match.
+        """
+
+        values = []
+        if values_2 is not None or values_1 is not None:
+            if values_2 is None:
+                values = values_1
+            elif values_1 is None:
+                values = values_2
+            else:
+                for (val_1, val_2) in zip(values_1, values_2):
+                    if len(val_1) != len(val_2):
+                        raise ValueError(
+                            f"The number of left parameters (currently {len(val_1)})"
+                            f"has to be equal to the number of right parameters."
+                            f"(currently {len(val_2)})"
+                        )
+                    values.append(val_1 + val_2)
+
+        # set values
+        self._parameter_values = values
+
+    def _preprocess_inputs(
+        self,
+        circuits_1: Sequence[QuantumCircuit],
+        circuits_2: Sequence[QuantumCircuit],
+        values_1: Sequence[Sequence[float]] | None = None,
+        values_2: Sequence[Sequence[float]] | None = None,
+    ) -> None:
+        """Preprocess input circuits and parameter values and update corresponding lists.
+
+        Args:
+           circuits_1: (Parametrized) quantum circuits preparing the first list of quantum states.
+           circuits_2: (Parametrized) quantum circuits preparing the second list of quantum states.
+           values_1: Numerical parameters to be bound to the first circuits.
+           values_2: Numerical parameters to be bound to the second circuits.
+
+        """
+
+        self._set_circuits(circuits_1, circuits_2)
+
+        values_1 = self._check_values_shape(values_1, circuits_1, "1")
+        values_2 = self._check_values_shape(values_2, circuits_2, "2")
+
+        self._set_values(values_1, values_2)
+
```

Comment:
In regards of my comment above about the interface being defined here and run() is there any reason not to move run() from the Fidelity to here. It seems generic enough defined over self._run() which is abstract here and will require a concrete class to implement.